source/Cosmos.System2/Audio/AudioMixer.cs

```diff
@@ -0,0 +1,177 @@
+            if(mixBuffer == null || mixBuffer.Size != buffer.Size || mixBuffer.format != buffer.format) {
+                mixBuffer = new AudioBuffer(buffer.Size, buffer.format);
+                streamReader = new AudioBufferReader(mixBuffer);
+            }
+
+            if(cachedBuffer != buffer)
+            {
+                cachedBuffer = buffer;
+                cachedOutputWriter = new(buffer, new SampleFormat(AudioBitDepth.Bits16, 1, true));
+                cachedOutputReader = new(buffer);
+            }
+
+            if (Streams.Count == 1)
+            {
+                AudioStream stream = Streams[0];
+
+                if (stream.Depleted)
+                {
+                    Streams.Clear();
+                    buffer.Flush();
+                }
+                else
+                {
+                    Streams[0].Read(buffer);
+                }
+            }
+            else
+            {
+                buffer.Flush();
+
+                // Mix all streams together
+                // A reverse for-loop so we can modify the list while iterating
+                for (int i = Streams.Count - 1; i >= 0; i--)
+                {
+                    AudioStream stream = Streams[i];
+
+                    // Do not process streams with no data
+                    if (stream.Depleted)
+                    {
+                        Streams.RemoveAt(i);
+                        continue;
+                    }
+
+                    // Read the stream to our mixing buffer - this will also
+                    // convert it to the right format
+                    stream.Read(mixBuffer);
+
+                    // Iterate through every sample
+                    for (int j = 0; j < mixBuffer.Size; j++)
+                    {
+                        // Iterate through each channel of the output buffer
+                        // Our mixing buffer has the same amount of channels as the output buffer
+                        for (int channel = 0; channel < buffer.format.channels; channel++)
+                        {
+                            short sum = SaturationAdd(
+                                cachedOutputReader.ReadChannelInt16(j, channel),
+                                streamReader.ReadChannelInt16(j, channel)
+                            );
+
+                            cachedOutputWriter.WriteChannel(
+                                (byte*)&sum,
+                                j, channel
+                            );
+                        }
+                    }
+                }
+            }
+
+            pos += (uint)buffer.Size;
+
+            ApplyPostProcessing(buffer);
+        }
+
+        private static short SaturationAdd(short a, short b)
+        {
+            if(a > 0) {
+                if (b > short.MaxValue - a)
+                    return short.MaxValue;
+            }
+            else if (b < short.MinValue - a)
```

Comment:
Does this not cause an underflow?

Reply:
Yes, the method relies on detecting underflows/overflows in order to perform saturation arithmetic - you can test the code on a website like rextester.com

Reply:
In this else, `a <= 0`, so it doesn't overflow.

Reply:
We already check for this in the `if(a > 0)` statement

Reply:
Thanks, missed the effect of the `a > 0` check.